# Hw2_Lundstrum

# This homework is due on 9/22 Tuesday at 11:59 pm

### Instructions
1) Start the first cell with your last name, HW/Lab number, etc. For example, HW2_solutions_basnet.

2) After you finish the HW, click File on the top left and click "download as." You will see many options. Please save the file as .ipynb and .html file.

3) To submit HW solutions, go to our course in Canvas --> Assignments --> HW 2 --> Submit Assignment --> Click "Chose file" --> choose HTML file you saved --> Click another file--> choose ipynb file--submit. Once you submit, download  and check to make sure that you have submitted the correct files.
##### ( Both HTML and ipynb files are required for your HW to be graded from now on)

4) Your submission will be compared against other student's solutions and online databases to check plagiarism.

5) You can use this ipynb file to start your solution

6) I will make HWs due on Tuesdays so that you do not have 2 assignments due on Fridays. Each problem is worth 10 points.

In [1]:
import pandas as pd
import numpy as np
coluns2use = ['author_average_rating', 'author_gender','author_name','birthplace','book_average_rating',
              'book_title', 'genre_1','num_ratings', 'num_reviews', 'pages','score']

### Problem 1
Upload the grbooks.csv data in your Jupyter Notebook.  Then, select only the variables listed in the list columns2use in the previous cell and create a new data frame. Find the data type of each of the variables. Do all the variables have correct labels as objects (Categorical) or numbers? Which variable has a wrong datatype?

In [2]:
grb = pd.read_csv('grbooks.csv', decimal = ',')
grbspec = grb[coluns2use].copy()
grbspec.head()

author_average_rating author_gender            author_name  \
0                  4.01        female    Victoria Thompson\n   
1                  4.15          male        Stieg Larsson\n   
2                     4        female  Mimi Jean Pamfiloff\n   
3                  3.88          male          José Donoso\n   
4                   4.1        female    Patricia C. Wrede\n   

             birthplace book_average_rating  \
0   United States\n                    4.02   
1          Sweden\n                    4.13   
2     United States\n                  3.99   
3           Chile\n                    4.14   
4   United States\n                    4.01   

                                          book_title  genre_1  num_ratings  \
0               \n      Murder on St. Mark's Place\n  Mystery         5260   
1          \n      The Girl with the Dragon Tattoo\n  Fiction      2229163   
2                     \n      Tailored for Trouble\n  Romance         2151   
3                \n      The Obscene Bird of Night\n  Fiction         1844   
4  \n      Sorcery & Cecelia: or The Enchanted Ch...  Fantasy        17051   

   num_reviews pages  score  
0          375   277   3230  
1        65227   465   3062  
2          391   354   4585  
3          173   438   1533  
4         1890   326   2105

In [3]:
grbspec.dtypes

author_average_rating    object
author_gender            object
author_name              object
birthplace               object
book_average_rating      object
book_title               object
genre_1                  object
num_ratings               int64
num_reviews               int64
pages                    object
score                     int64
dtype: object

No, author_average_rating, book_average_rating, and pages are all objects when they should be numbers.

In [4]:
grbspec['author_average_rating'] = grbspec['author_average_rating'].astype(float)
grbspec['book_average_rating'] = grbspec['book_average_rating'].astype(float)
grbspec['pages'] = pd.to_numeric(grbspec['pages'], errors='coerce')
grbspec.dtypes

author_average_rating    float64
author_gender             object
author_name               object
birthplace                object
book_average_rating      float64
book_title                object
genre_1                   object
num_ratings                int64
num_reviews                int64
pages                    float64
score                      int64
dtype: object

### Problem 2
Continuing with the data you obtained in problem 1, find the $\textbf{median}$ of all the numerical variables you saw in problem 1. Write a paragraph commenting on the median of these variables. 

In [5]:
grbspec.describe()

author_average_rating  book_average_rating   num_ratings  \
count           22891.000000         22891.000000  2.289100e+04   
mean                3.960368             3.951456  4.668349e+04   
std                 0.240421             0.291317  1.800698e+05   
min                 1.820000             0.000000  0.000000e+00   
25%                 3.810000             3.770000  8.200000e+02   
50%                 3.970000             3.960000  4.403000e+03   
75%                 4.120000             4.140000  2.014300e+04   
max                 5.000000             5.000000  3.820921e+06   

         num_reviews         pages          score  
count   22891.000000  22888.000000   22891.000000  
mean     2324.754794    332.996024    3893.028483  
std      6837.503574    219.329752   11022.018268  
min         0.000000      0.000000      55.000000  
25%       106.000000    237.000000     832.000000  
50%       384.000000    315.000000    1727.000000  
75%      1504.000000    384.000000    3597.500000  
max    147696.000000   6680.000000  598270.000000

The median author average rating is 3.97. The median book average rating is 3.96. The median number of ratings is 4400 ratings. The median number of reviews is 384 reviews. The median author score is 1727.

### Problem 3 
Find the $\textbf{mode}$ of all the numerical variables you saw in problem 1. Write a paragraph commenting on the mode of these variables. 

In [6]:
grbspec.mode(numeric_only=True,dropna=True)

author_average_rating  book_average_rating  num_ratings  num_reviews  \
0                   3.94                 3.96           44           17   

   pages  score  
0  320.0    200

The most common author average rating is 3.94. The most common average book rating is 3.96. The most common number of ratings is 44 ratings. The most common number of reviews is 17 reviews. The most common score is 200.

### Problem 4
A distribution (here it is one of the numerical columns) in which the values of mean, median, and mode coincide (i.e., mean = median = mode) is known as a normal distribution (a perfect data set).It has a bell shapped curve when plotted and is also symmetric about the mean. Conversely, when mean, median, and mode values are not equal, the distribution is known as asymmetrical or skewed distribution. In moderately skewed or asymmetrical distribution, a very important relationship exists among these three measures of central tendencies. In such distributions, the distance between the mean and median is about one-third of the mean and mode distance. Karl Pearson, a renowned statistician, expressed this relationship as:
$Mode = 3  \,\,Median - 2\,\, Mean $. 

![](mcube.png) 


Find $Mode - 3  \,\,Median + 2\,\, Mean $ for all numerical columns. Symmetric data is hard to come by. If this number is close to 0 for a column, the column data is close to being symmetric, and if it is far from 0 ( positive or negative), the column data is asymmetric. Which column is closest to being symmetric, and which one is the most asymmetric?

In [7]:
new = grbspec.mode(numeric_only=True, dropna=True)-(3*grbspec.median())+(2*grbspec.mean())

In [8]:
print(new)

   author_average_rating  book_average_rating  num_ratings  num_reviews  \
0              -0.049263            -0.017088  80201.97558  3514.509589   

       pages        score  
0  40.992048  2805.056966  


book_average_rating appears to be closest to being symmetric and num_ratings appears to be the most asymmetric.

### Problem 5
We can use sample function from pandas to randomly select only few  rows from our data frame. Sampling is very useful when the data is too big and not all of it can be handled in a computer.
More about sample function is explained here.
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html


Using the sample function, sample 1000 data points from the data you created in problem 1 and give it a new name. Now find the means  of this new sampled data and compare with the means of the data you sampled from. Which variable's mean  from this sampled data setis close to the mean of the original data? What does it imply when these two means are very close? 

In [9]:
randsamp = grbspec.sample(n=1000, random_state=1)

In [10]:
randsamp.describe()

author_average_rating  book_average_rating   num_ratings   num_reviews  \
count            1000.000000          1000.000000  1.000000e+03   1000.000000   
mean                3.959190             3.956930  4.590409e+04   2426.488000   
std                 0.243364             0.285465  1.567685e+05   6814.472764   
min                 2.970000             2.910000  7.000000e+00      0.000000   
25%                 3.817500             3.770000  8.952500e+02    111.250000   
50%                 3.960000             3.960000  4.971000e+03    419.000000   
75%                 4.120000             4.150000  2.306400e+04   1712.750000   
max                 4.950000             4.940000  2.034222e+06  98584.000000   

             pages          score  
count  1000.000000    1000.000000  
mean    352.384000    4399.978000  
std     320.312205   13894.269649  
min       0.000000      96.000000  
25%     240.000000     897.750000  
50%     320.000000    1804.000000  
75%     400.000000    3641.250000  
max    6576.000000  271577.000000

Several of the variables means are close to the means of the original data. author_average_rating has the closest mean to the original. The two means being very close implies that the random sample is a good representation of the population.

### Problem 6
Make a frequency table of the variable "genre_1". Your answer should look like a data frame with indices ( with no column name), and two other column names "genre_1" and "frequency". The following Panda functions will be useful: groubby, value_counts, reset_index() and to_frame(). https://stackoverflow.com/questions/39778686/pandas-reset-index-after-groupby-value-counts and https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reset_index.html

In [11]:
fgrb = grbspec.genre_1.value_counts()
print(fgrb)

Romance        3397
Fantasy        3397
Young Adult    2497
Fiction        2172
Historical     1399
               ... 
How To            1
Dc Comics         1
Design            1
Pulp              1
Love              1
Name: genre_1, Length: 124, dtype: int64


### Problem 7. 
Using the column  "score", to create a new column called new_score that will have a score out of 5. For this, apply the formula  $\text{new_score} = \dfrac{\text{score}}{\text{max(score)}}\times 5$. Print the description of this new colum using describe function.

In [12]:
grbspec["new_score"] = grbspec["score"]/ grbspec["score"].max() * 5
grbspec.describe()

author_average_rating  book_average_rating   num_ratings  \
count           22891.000000         22891.000000  2.289100e+04   
mean                3.960368             3.951456  4.668349e+04   
std                 0.240421             0.291317  1.800698e+05   
min                 1.820000             0.000000  0.000000e+00   
25%                 3.810000             3.770000  8.200000e+02   
50%                 3.970000             3.960000  4.403000e+03   
75%                 4.120000             4.140000  2.014300e+04   
max                 5.000000             5.000000  3.820921e+06   

         num_reviews         pages          score     new_score  
count   22891.000000  22888.000000   22891.000000  22891.000000  
mean     2324.754794    332.996024    3893.028483      0.032536  
std      6837.503574    219.329752   11022.018268      0.092116  
min         0.000000      0.000000      55.000000      0.000460  
25%       106.000000    237.000000     832.000000      0.006953  
50%       384.000000    315.000000    1727.000000      0.014433  
75%      1504.000000    384.000000    3597.500000      0.030066  
max    147696.000000   6680.000000  598270.000000      5.000000

### Problem 8.
We have now 3 columns that represent ratings out of 5, namely author_average_rating, book_average_rating, and a new_score we just created in problem 7. Let's create a column called super_score which takes into consideration of all the 3 columns with rating scores out of 5. Give 24%, 34% and 42% weights for the columns author_average_rating, book_average_rating, and a new_score respectively to create the values for super_score column that is also out of 5. Now use the describe function to show the details of this new column.

In [13]:
grbspec["super_score"] = (grbspec["author_average_rating"]*.24)+(grbspec["book_average_rating"]*.34)+(grbspec["new_score"]*.42)
grbspec.describe()

author_average_rating  book_average_rating   num_ratings  \
count           22891.000000         22891.000000  2.289100e+04   
mean                3.960368             3.951456  4.668349e+04   
std                 0.240421             0.291317  1.800698e+05   
min                 1.820000             0.000000  0.000000e+00   
25%                 3.810000             3.770000  8.200000e+02   
50%                 3.970000             3.960000  4.403000e+03   
75%                 4.120000             4.140000  2.014300e+04   
max                 5.000000             5.000000  3.820921e+06   

         num_reviews         pages          score     new_score   super_score  
count   22891.000000  22888.000000   22891.000000  22891.000000  22891.000000  
mean     2324.754794    332.996024    3893.028483      0.032536      2.307648  
std      6837.503574    219.329752   11022.018268      0.092116      0.156127  
min         0.000000      0.000000      55.000000      0.000460      0.871881  
25%       106.000000    237.000000     832.000000      0.006953      2.211681  
50%       384.000000    315.000000    1727.000000      0.014433      2.310253  
75%      1504.000000    384.000000    3597.500000      0.030066      2.406540  
max    147696.000000   6680.000000  598270.000000      5.000000      4.625000

### Problem 9
Use super_score column to find out how many unique authors have a super_score of 3.9 or more. Let's call these authors, super_authors. Create a dataframe called super_authors that has the following columns: ['author_name','author_gender','birthplace','book_title'] and show the head of the data.

In [14]:
columns2use = ['author_name', 'author_gender', 'birthplace', 'book_title']
super_authors = grbspec.loc[(grbspec["super_score"]>=3.9)]
super_authors = super_authors[columns2use].copy()
super_authors.head()

author_name author_gender             birthplace  \
919       Oscar Wilde\n          male          Ireland\n       
920     Lewis Carroll\n          male   United Kingdom\n       
921  Shel Silverstein\n          male    United States\n       

                                            book_title  
919               \n      The Picture of Dorian Gray\n  
920  \n      Alice's Adventures in Wonderland & Thr...  
921                          \n      The Giving Tree\n

### Problem 10
Save the dataframe super_authors you created in problem 9 to your working folder (where you have this HW saved) as a csv file and read it back to Jupyter Notebook and print the head of the data.

In [15]:
super_authors.to_csv(r'super_authors.csv', index = False, header = True)

In [16]:
super_authors2 = pd.read_csv('super_authors.csv', decimal = ',')
super_authors2.head()

author_name author_gender             birthplace  \
0       Oscar Wilde\n          male          Ireland\n       
1     Lewis Carroll\n          male   United Kingdom\n       
2  Shel Silverstein\n          male    United States\n       

                                          book_title  
0               \n      The Picture of Dorian Gray\n  
1  \n      Alice's Adventures in Wonderland & Thr...  
2                          \n      The Giving Tree\n

## The following 2 problems are extra credit problems and are optional. You won't get any help on these problems.

### Bonus Problem 1 (Bonus 10 points)
In super_author data set, there are a few columns that have entries with $\text{\n}$ notation. Find these columns and remove $\text{\n}$ from all the column entries using regular expressions.  Read more about handling strings and regular expressions here. https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html, https://www.geeksforgeeks.org/replace-values-in-pandas-dataframe-using-regex/
, and here https://docs.python.org/2/library/re.html.

In [17]:
super_authorsnon = super_authors.replace(to_replace = '\n', value = '', regex = True)

In [18]:
super_authorsnon.head()

author_name author_gender           birthplace  \
919       Oscar Wilde          male          Ireland       
920     Lewis Carroll          male   United Kingdom       
921  Shel Silverstein          male    United States       

                                            book_title  
919                         The Picture of Dorian Gray  
920        Alice's Adventures in Wonderland & Throu...  
921                                    The Giving Tree

### Bonus Problem 2 (Bonus 10 points)
You may have noticed that the column named  "pages" from the data set in problem 1 seem to have numerical looking values but it is not showing up in our data summary. This is because this columns appears to be a string in our data set. Data values in this columns are just  the total number of pages in each book and these should be numbers not strings. Change this column from string to numeric and find mean, median and mode for this column. If the error comes in the process of converting to numerical, find out what the error is and fix it.

I already did in problem 1, for reference the code I used was grbspec['pages'] = pd.to_numeric(grbspec['pages'], errors='coerce')

In [19]:
grbspec.describe()

author_average_rating  book_average_rating   num_ratings  \
count           22891.000000         22891.000000  2.289100e+04   
mean                3.960368             3.951456  4.668349e+04   
std                 0.240421             0.291317  1.800698e+05   
min                 1.820000             0.000000  0.000000e+00   
25%                 3.810000             3.770000  8.200000e+02   
50%                 3.970000             3.960000  4.403000e+03   
75%                 4.120000             4.140000  2.014300e+04   
max                 5.000000             5.000000  3.820921e+06   

         num_reviews         pages          score     new_score   super_score  
count   22891.000000  22888.000000   22891.000000  22891.000000  22891.000000  
mean     2324.754794    332.996024    3893.028483      0.032536      2.307648  
std      6837.503574    219.329752   11022.018268      0.092116      0.156127  
min         0.000000      0.000000      55.000000      0.000460      0.871881  
25%       106.000000    237.000000     832.000000      0.006953      2.211681  
50%       384.000000    315.000000    1727.000000      0.014433      2.310253  
75%      1504.000000    384.000000    3597.500000      0.030066      2.406540  
max    147696.000000   6680.000000  598270.000000      5.000000      4.625000